### PR TITLE
KRB5_TRACE documentation for gssproxy

### PR DESCRIPTION
**Pull request by jacobshivers**
Added KRB5_TRACE.md to provide a brief overview on how to enable
low level Kerberos tracing below gssapi via KRB5_TRACE.

Signed-off-by: Jacob Shivers <jacob.shivers@redhat.com>

### CODE REVIEW

**Comment by simo5** [comment]
@jacobshivers I am ok with this, but I am curious, KRB5_TRACE output will already be logged to the journal at debug level 3, why do you need to override the output and send it to a file? Is this to avoid flooding the regular logs with debug information?

**Comment by jacobshivers** [comment]
In my tests, setting debug_level = 3 does not sent trace information to the system journal. I had to manually specific a location for KRB5_TRACE for the data to be logged.

**Comment by simo5** [comment]
Well this line:
https://github.com/gssapi/gssproxy/blob/dbc4820757cc782c6387a843053d133ebdf8600e/src/gp_debug.c#L18
sends KRB5_TRACE to stderr, so it should definitely end up in the journal.
If not I would know why, did you modify your systemd unit in some way?

Or maybe there have been changes to what systemd logs now from stdout/stderr?

**Comment by simo5** [comment]
Also on my system:
$ systemctl show gssproxy.service |grep Standard
StandardInput=null
StandardOutput=journal
StandardError=inherit

What does your system show?

**Comment by simo5** [comment]
> StandardError=inherit
and does adding StandardError=journal in a systemd drop file make any difference?

File: docs/KRB5_TRACE.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+Enabling `KRB5_TRACE` output as described in `# man gssproxy.conf`:
+
+**At level 3 and above, KRB5_TRACE output is logged. If KRB5_TRACE was already set in the execution environment, trace output is sent to its value instead.**
+
+can be done either by including `KRB5_TRACE=/path/to/location` when executing gssproxy at the command line or by including a location within a drop file loaded by systemd.
+
+The output can be directed to any location, but gssproxy only has write access to `/var/lib/gssproxy` by default. This means that for a host system using SELinux either a custom module policy will need to be created or SELinux will need to be put into permissive mode.
+
+As setting `KRB5_TRACE` output is not designed to be continually logged nor treated as traditional log output, it is recommended to direct the trace output to `/var/lib/gssproxy`.
```
Comment (simo5) [inline docs/KRB5_TRACE.md:11]: Once we figure out why you do not get KRB5_TRACE output in system log I would like to change this "recommnded" to a less "official sounding" verbiage.
Something like "if you do not want to send all this debug info to the system log, you may elect to send it to a file in var/lib/gssproxy by ..."

**Review by simo5 — CHANGES_REQUESTED** [review]

**Comment by jacobshivers** [comment]
RHEL 8.5
# systemctl show gssproxy | grep Standard
StandardInput=null
StandardInputData=
StandardOutput=journal
StandardError=inherit

Fedora 35:
# systemctl show gssproxy | grep Standard
StandardInput=null
StandardOutput=journal
StandardError=inherit

RHEL 7,9
# systemctl show gssproxy | grep Standard
StandardInput=null
StandardOutput=journal
StandardError=inherit

None of those systems direct KRB5_TRACE to the system journal by default.

Adding

StandardError=journal requires adjusting max message size for rsyslogd:

# journalctl -b0 | grep 'rsyslogd.* too long' | awk '{for (i=4;i<=NF;i++){printf "%s ",$i ; if (i==NF) print ""}}'
rhel-85.ad.tob.com rsyslogd[1085]: message too long (9523) with configured size 8096, begin of message is: GSSX_RES_INDICATE_MECHS( status: { 0 <None> 0 "" "" [ ] } mechs: [ { { 1 2 [v8.2102.0-5.el8 try https://www.rsyslog.com/e/2445 ] 
rhel-85.ad.tob.com rsyslogd[1085]: message too long (19205) with configured size 8096, begin of message is: GSSX_ARG_ACQUIRE_CRED( call_ctx: { "" [ ] } input_cred_handle: { { "picard@ [v8.2102.0-5.el8 try https://www.rsyslog.com/e/2445 ] 
rhel-85.ad.tob.com rsyslogd[1085]: message too long (18732) with configured size 8096, begin of message is: GSSX_RES_ACQUIRE_CRED( status: { 0 { 1 2 840 113554 1 2 2 } 0 "" "" [ ] } o [v8.2102.0-5.el8 try https://www.rsyslog.com/e/2445 ] 
rhel-85.ad.tob.com rsyslogd[1085]: message too long (19088) with configured size 8096, begin of message is: GSSX_ARG_INIT_SEC_CONTEXT( call_ctx: { "" [ ] } context_handle: <Null> cred [v8.2102.0-5.el8 try https://www.rsyslog.com/e/2445 ] 
rhel-85.ad.tob.com rsyslogd[1085]: message too long (9272) with configured size 8096, begin of message is: GSSX_RES_INIT_SEC_CONTEXT( status: { 1 { 1 2 840 113554 1 2 2 } 0 "The routi [v8.2102.0-5.el8 try https://www.rsyslog.com/e/2445 ] 
rhel-85.ad.tob.com rsyslogd[1085]: message too long (21872) with configured size 8096, begin of message is: GSSX_ARG_INIT_SEC_CONTEXT( call_ctx: { "" [ ] } context_handle: { [ 00092af [v8.2102.0-5.el8 try https://www.rsyslog.com/e/2445 ]


# grep Max /etc/rsyslog.conf
$MaxMessageSize 24k

Even after increasing the max message size, I still do not see KRB5_TRACE messages

**Comment by jacobshivers** [comment]
And ack to adjusting language. It is why I also avoided even providing a description for a custom SELinux module. That can could lead to some unnecessary changes.

**Comment by jacobshivers** [comment]
It is an issue with SELinux

~~~~
# systemctl show gssproxy | grep Standard
StandardInput=null
StandardInputData=
StandardOutput=journal
StandardError=journal

# ausearch  -m avc -ts today
<no matches>

# mount 8x1-nfs.ad.tob.com:/exports/secure /mnt -o sec=krb5 -vvv
mount.nfs: timeout set for Mon Mar  7 17:13:12 2022
mount.nfs: trying text-based options 'sec=krb5,vers=4.2,addr=192.168.133.140,clientaddr=192.168.133.162'

# ls -l /mnt
total 4
-rw-r--r--. 1 f35$        domain users   0 Mar  3 21:36 bleh
-rw-r--r--. 1 data        domain users   0 May  4  2021 data
-rw-r--r--. 1 data        domain users   0 May  4  2021 foo
-rw-r--r--. 1 picard      domain users   0 May  4  2021 picard
-rw-r--r--. 1 8x1-client$ domain users   0 Oct  5 15:06 root
-rw-r--r--. 1 nobody      domain users   0 Jan 19 18:32 test_sub

# su - picard

$ ls -l /mnt
total 4
-rw-r--r--. 1 f35$        domain users   0 Mar  3 21:36 bleh
-rw-r--r--. 1 data        domain users   0 May  4  2021 data
-rw-r--r--. 1 data        domain users   0 May  4  2021 foo
-rw-r--r--. 1 picard      domain users   0 May  4  2021 picard
-rw-r--r--. 1 8x1-client$ domain users   0 Oct  5 15:06 root
-rw-r--r--. 1 nobody      domain users   0 Jan 19 18:32 test_sub

# ausearch  -m avc -ts today
----
time->Mon Mar  7 17:11:12 2022
type=PROCTITLE msg=audit(1646691072.753:95): proctitle=2F7573722F7362696E2F67737370726F7879002D44
type=PATH msg=audit(1646691072.753:95): item=1 name="/dev/stderr" inode=23449 dev=00:09 mode=0140777 ouid=0 ogid=0 rdev=00:00 obj=system_u:system_r:init_t:s0 nametype=NORMAL cap_fp=0 cap_fi=0 cap_fe=0 cap_fver=0 cap_frootid=0
type=PATH msg=audit(1646691072.753:95): item=0 name="/dev/stderr" inode=29361 dev=00:05 mode=0120700 ouid=0 ogid=0 rdev=00:00 obj=system_u:system_r:gssproxy_t:s0 nametype=NORMAL cap_fp=0 cap_fi=0 cap_fe=0 cap_fver=0 cap_frootid=0
type=CWD msg=audit(1646691072.753:95): cwd="/"
type=SYSCALL msg=audit(1646691072.753:95): arch=c000003e syscall=257 success=no exit=-6 a0=ffffff9c a1=55b56e24928b a2=441 a3=180 items=2 ppid=1 pid=862 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="gssproxy" exe="/usr/sbin/gssproxy" subj=system_u:system_r:gssproxy_t:s0 key=(null)
type=AVC msg=audit(1646691072.753:95): avc:  denied  { write } for  pid=862 comm="gssproxy" name="fd" dev="proc" ino=29360 scontext=system_u:system_r:gssproxy_t:s0 tcontext=system_u:system_r:gssproxy_t:s0 tclass=dir permissive=0
type=AVC msg=audit(1646691072.753:95): avc:  denied  { dac_override } for  pid=862 comm="gssproxy" capability=1  scontext=system_u:system_r:gssproxy_t:s0 tcontext=system_u:system_r:gssproxy_t:s0 tclass=capability permissive=0

~~~~

**Comment by jacobshivers** [comment]
Actually, even setting SELinux into permissive did not resolve this nor did setting StandardError output to a fil:

~~~~
# systemctl show gssproxy | grep Standard
StandardInput=null
StandardInputData=
StandardOutput=journal
StandardError=file

# systemctl cat gssproxy | grep Error
StandardError=file:/var/lib/gssproxy/gssproxy.krb5_trace

# mount 8x1-nfs.ad.tob.com:/exports/secure /mnt -o sec=krb5 -vvv
mount.nfs: timeout set for Mon Mar  7 17:29:03 2022
mount.nfs: trying text-based options 'sec=krb5,vers=4.2,addr=192.168.133.140,clientaddr=192.168.133.162'

# ls -l /mnt
total 4
-rw-r--r--. 1 f35$        domain users   0 Mar  3 21:36 bleh
-rw-r--r--. 1 data        domain users   0 May  4  2021 data
-rw-r--r--. 1 data        domain users   0 May  4  2021 foo
-rw-r--r--. 1 picard      domain users   0 May  4  2021 picard
-rw-r--r--. 1 8x1-client$ domain users   0 Oct  5 15:06 root
-rw-r--r--. 1 nobody      domain users   0 Jan 19 18:32 test_sub

# su - picard
Last login: Mon Mar  7 17:25:09 EST 2022 on pts/0
[picard@rhel-85 ~]$ ls -l /mnt
total 4
-rw-r--r--. 1 f35$        domain users   0 Mar  3 21:36 bleh
-rw-r--r--. 1 data        domain users   0 May  4  2021 data
-rw-r--r--. 1 data        domain users   0 May  4  2021 foo
-rw-r--r--. 1 picard      domain users   0 May  4  2021 picard  
-rw-r--r--. 1 8x1-client$ domain users   0 Oct  5 15:06 root
-rw-r--r--. 1 nobody      domain users   0 Jan 19 18:32 test_sub

$ exit

# grep -c TGT /var/lib/gssproxy/gssproxy.krb5_trace 
0

# ausearch -m avc -ts today -i
----
type=PROCTITLE msg=audit(03/07/2022 17:26:40.253:134) : proctitle=(gssproxy)
type=PATH msg=audit(03/07/2022 17:26:40.253:134) : item=3 name=(null) inode=1465128 dev=fd:00 mode=file,644 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:gssproxy_var_lib_t:s0 nametype=CREATE cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=PATH msg=audit(03/07/2022 17:26:40.253:134) : item=2 name=(null) inode=1376645 dev=fd:00 mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:gssproxy_var_lib_t:s0 nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=PATH msg=audit(03/07/2022 17:26:40.253:134) : item=1 name=(null) nametype=CREATE cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=PATH msg=audit(03/07/2022 17:26:40.253:134) : item=0 name=(null) inode=1376645 dev=fd:00 mode=dir,755 ouid=root ogid=root rdev=00:00 obj=system_u:object_r:gssproxy_var_lib_t:s0 nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0
type=CWD msg=audit(03/07/2022 17:26:40.253:134) : cwd=/
type=SYSCALL msg=audit(03/07/2022 17:26:40.253:134) : arch=x86_64 syscall=openat success=yes exit=3 a0=0xffffff9c a1=0x55e9caf1a390 a2=O_WRONLY|O_CREAT|O_NOCTTY a3=0x1a4 items=4 ppid=1 pid=1814 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=(gssproxy) exe=/usr/lib/systemd/systemd subj=system_u:system_r:init_t:s0 key=(null)
type=AVC msg=audit(03/07/2022 17:26:40.253:134) : avc:  denied  { write open } for  pid=1814 comm=(gssproxy) path=/var/lib/gssproxy/gssproxy.krb5_trace dev="dm-0" ino=1465128 scontext=system_u:system_r:init_t:s0 tcontext=system_u:object_r:gssproxy_var_lib_t:s0 tclass=file permissive=1
type=AVC msg=audit(03/07/2022 17:26:40.253:134) : avc:  denied  { create } for  pid=1814 comm=(gssproxy) name=gssproxy.krb5_trace scontext=system_u:system_r:init_t:s0 tcontext=system_u:object_r:gssproxy_var_lib_t:s0 tclass=file permissive=1
type=AVC msg=audit(03/07/2022 17:26:40.253:134) : avc:  denied  { add_name } for  pid=1814 comm=(gssproxy) name=gssproxy.krb5_trace scontext=system_u:system_r:init_t:s0 tcontext=system_u:object_r:gssproxy_var_lib_t:s0 tclass=dir permissive=1
~~~~

So once the issue with writing to stderr is resolved, there will probably need to be some adjustments for SELinux as well.

**Comment by simo5** [comment]
This is so odd ..

**Comment by jacobshivers** [comment]
I have the same label

~~~~
# ps Z -C gssproxy
LABEL                               PID TTY      STAT   TIME COMMAND
system_u:system_r:gssproxy_t:s0    1932 ?        Ssl    0:00 /usr/sbin/gssproxy -D

# ls -lZ /usr/sbin/gssproxy
-rwxr-xr-x. 1 root root system_u:object_r:gssproxy_exec_t:s0 135480 Oct 29  2020 /usr/sbin/gssproxy

# rpm -qf $(which gssproxy)
gssproxy-0.8.0-19.el8.x86_64

~~~~

**Comment by simo5** [comment]
Ok then it is the attempt to:
StandardError=file:/var/lib/gssproxy/gssproxy.krb5_trace that is causing them and it is not really relevant.

The other two avcs I see them too but one is related to "proc" and the other is unclear but seems irrelevant.

I also verified that setting debug_level=3 does not get me the KRB5_TRACE output.

I suspect the problem is that opening /dev/stderr explicitly does not work because the journal is not reading from the devices, but rather creating new files for FD 0, 1, 2, so opening /dev/stderr will work on the command line when manually running gssproxy but does not work to pipe to the journal.

Could you change your configuration back to pristine and then only add:
`Environment="KRB5_TRACE=/var/lib/gssproxy/gssproxy.krb5_trace"`
as a drop file.

If this works I think the readme text should note that KRB5_TRACE does no work in a systemd environment by just setting debug_level= 3 and provide this Environment variable setting trick as the solution to get out krb5 trace logs.

**Comment by simo5** [comment]
I think all should be in the new doc, I am also ok with a mention in the manpage that tells people to look at the KRB5_TRACE.md for more info

**Review by jacobshivers — COMMENTED** [review]
Proposed addition detailing systemd behavior.

File: man/gssproxy.conf.5.xml [hunk 2]
```diff
@@ -216,7 +216,7 @@
                             At level 3 and above, KRB5_TRACE output is logged.
                             If KRB5_TRACE was already set in the execution
                             environment, trace output is sent to its value
-                            instead.
+				instead. Refer to docs/KRB5_TRACE.md for more information.
```
Comment (simo5) [inline man/gssproxy.conf.5.xml:219]: Looks like indentation issue here.

File: docs/KRB5_TRACE.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
+# Setting KRB5_TRACE for gssproxy
+
+Enabling `KRB5_TRACE` output as described in `# man gssproxy.conf`:
+
+**At level 3 and above, KRB5_TRACE output is logged. If KRB5_TRACE was already set in the execution environment, trace output is sent to its value instead.**
+
+can be done either by including `KRB5_TRACE=/path/to/location` when executing gssproxy at the command line or by including a location within a drop file loaded by systemd. Systemd does not open `/dev/stderr` and instead creates a file descriptor for stderr when gssproxy is started. A drop file is necessary to record `KRB5_TRACE` information as gssproxy directs `KRB5_TRACE` to `/dev/stderr` at this time.
```
Comment (simo5) [inline docs/KRB5_TRACE.md:7]: Replace this:
`Systemd does not open `/dev/stderr` and instead creates a file descriptor for stderr when gssproxy is started.`
with
`Systemd uses unix sockets to redirect outputs to the journal, and this means /dev/stderr ends up pointing to a name that cannot be used as a path to open the stderr descriptor`

And later "at this time" -> "by default"

**Review by simo5 — CHANGES_REQUESTED** [review]
I think with these two changes we can merge this

File: man/gssproxy.conf.5.xml (hunk 2 above)
Comment (jacobshivers) [reply man/gssproxy.conf.5.xml]: When I had added a line it did the weird indent and I did not go back and fix. Corrected now, but it seems the 'i' is a space too far to the left in `$ git show`.

I was not too sure the best way to preview the XML file outside of a text editor to make sure that it was actually structured appropriately.

Comment (simo5) [reply man/gssproxy.conf.5.xml]: not a big deal

File: docs/KRB5_TRACE.md (hunk 3 above)
Comment (jacobshivers) [reply docs/KRB5_TRACE.md]: Done. I had only mentioned "at this time" due to a potential change in how KRB5_TRACE is handled, but "by default" is more clear.

**Pull request merged**